language/move-symbol-pool/src/pool.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! A pool of uniqued string data akin to a hash set.
+//!
+//! The string pool is represented as a statically sized contiguous array of
+//! buckets. Each bucket represents space within which an entry in the pool
+//! may be allocated. An entry is an intrusively-linked list of with two data:
+//! a heap-allocated string, and the hash value of that string data.
+//!
+//! Insertion performance into the pool is comparable to a hash set, because the
+//! underlying data structure is similar: a hashing function determines which
+//! bucket in the array the string ought to be inserted into. The entries in
+//! that bucket are iterated upon. If no entries match the string, a new
+//! entry is appended to the end of the linked list of entries.
+//!
+//! * Why not use a [`HashSet`]? A set is dynamically resized as elements are
+//!   added. Ideally converting a [`Symbol`] to its string value is as
+//!   performant as dereferencing a pointer. But implementing the [`Symbol]` as
+//!   a pointer would not be safe if the data being pointed to could be
+//!   reallocated.
+//! * Why not use a [`LinkedList`]? A linked list does not unique the elements.
+//!   Ensuring the elements in the list are unique would require traversing the
+//!   list, which would not be performant for large lists.
+//! * Why not use a [`HashSet`] in conjunction with a [`LinkedList`]? This would
+//!   be simpler to implement, but would involve storing the string data twice:
+//!   once in the set for uniqueness checking, and once in the linked list to
+//!   maintain a constant memory address for the string data. This pool
+//!   implementation is more space-efficient.
+
+use std::{
+    collections::hash_map::DefaultHasher,
+    hash::{Hash, Hasher},
+    ptr::NonNull,
+};
+
+/// The number of buckets in the pool's contiguous array.
+const NB_BUCKETS: usize = 1 << 12; // 4096
+/// A mask used to convert a string hash into the index of one of the contiguous buckets.
+const BUCKET_MASK: u64 = NB_BUCKETS as u64 - 1;
+
+/// A bucket is a space on the heap within which an entry may be allocated.
+type Bucket = Option<Box<Entry>>;
+
+/// A string in the pool.
+pub(crate) struct Entry {
+    pub(crate) string: Box<str>,
+    hash: u64,
+    next: Bucket,
+}
+
+/// A contiguous array of buckets.
+pub(crate) struct Pool(Box<[Bucket; NB_BUCKETS]>);
+
+impl Pool {
+    /// Allocates a contiguous array of buckets on the heap. As strings are
+    /// inserted into the pool, buckets in this array are filled with an entry.
+    pub(crate) fn new() -> Self {
+        let vec = std::mem::ManuallyDrop::new(vec![0_usize; NB_BUCKETS]);
+        Self(unsafe { Box::from_raw(vec.as_ptr() as *mut [Bucket; NB_BUCKETS]) })
```

Comment:
You can avoid the unsafe here with [into_boxed_slice](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.into_boxed_slice):
```suggestion
        Self(vec![None; NB_BUCKETS].into_boxed_slice())
```

Reply:
nvm i didn't realize you wanted a `Box<[Bucket; NB_BUCKETS]>`